### PR TITLE
added estimateSmartFee rpc

### PR DESCRIPTION
**Pull request by thedoublejay**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
Implemented RPC for #48 
1. mining.estimatesmartfee
#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

File: packages/jellyfish-api-core/src/category/mining.ts [hunk 1]
```diff
@@ -86,3 +98,12 @@ export interface MasternodeInfo {
   mintedblocks?: number
   lastblockcreationattempt?: string
 }
+
+/**
+ * @description return type of rpc `estimatesmartfee`
+ */
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/mining.ts:104]: don't need to specific this, everything in this package does that.

File: packages/jellyfish-api-core/src/category/mining.ts [hunk 2]
```diff
@@ -37,6 +37,16 @@ export class Mining {
...
     return await this.client.call('getmininginfo', [], 'number')
   }
+
+  /**
+   *
+   * @param {number} confirmationTarget in blocks (1 - 1008)
+   * @param {EstimateMode} [estimateMode='CONSERVATIVE'] estimateMode of fees.
+   * @returns {Promise<SmartFeeEstimation>}
+   */
+  async estimateSmartFee (confirmationTarget: number, estimateMode: wallet.Mode = wallet.Mode.CONSERVATIVE): Promise<SmartFeeEstimation> {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/mining.ts:47]: don't import `wallet` even though it's applicable. This is to prevent code smell where everything starts to depend on each other, making it hard to migrate around. For abstraction, yes, but for implementation no. `wallet` is an implementation.

**Review by fuxingloh — CHANGES_REQUESTED** [review]

**Comment by fuxingloh** [comment]
don't use semantic convention either, just stick to PR purpose

File: packages/jellyfish-api-core/src/category/mining.ts (hunk 2 above)
Comment (thedoublejay) [reply packages/jellyfish-api-core/src/category/mining.ts:47]: ok will update it

File: packages/jellyfish-api-core/__tests__/category/mining.test.ts [hunk 3]
```diff
@@ -40,6 +49,8 @@ describe('masternode', () => {
   beforeAll(async () => {
     await container.start()
     await container.waitForReady()
+    await container.waitForWalletCoinbaseMaturity()
+    await container.waitForWalletBalanceGTE(100)
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/mining.test.ts:53]: you can revert this change, you don't need it, you have your own setup below

File: packages/jellyfish-api-core/src/category/mining.ts [hunk 4]
```diff
@@ -86,3 +96,11 @@ export interface MasternodeInfo {
...
   lastblockcreationattempt?: string
 }
+
+export interface SmartFeeEstimation {
+  feerate?: number
+  errors?: string[]
+  blocks: number
+}
+
+type EstimateMode = 'UNSET' | 'ECONOMICAL' | 'CONSERVATIVE'
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/mining.ts:106]: Currently, everyone put it above, we can follow that also, should use enum also. Enum we can add docs. plain types we can't

You can follow this but have it for yourself: 

https://github.com/DeFiCh/jellyfish/blob/ece0fdc6e24e2f9601fc4dfa474a9b6c924d507d/packages/jellyfish-api-core/src/category/wallet.ts#L3-L7

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/jellyfish-api-core/src/category/mining.ts (hunk 4 above)
Comment (thedoublejay) [reply packages/jellyfish-api-core/src/category/mining.ts:106]: Ok sure. I just remembered we had a discussion before about enums so I used types.

**Review by canonbrother — COMMENTED** [review]
other than fuxing's request.. overall look nice for me

File: packages/jellyfish-api-core/src/category/mining.ts (hunk 4 above)
Comment (fuxingloh) [reply packages/jellyfish-api-core/src/category/mining.ts]: I think you convinced me not to use type for everything. I only use `type` selectively now.

Comment (fuxingloh) [reply packages/jellyfish-api-core/src/category/mining.ts]: Also because of this, enum has JSDoc support but for type there isn't: https://github.com/DeFiCh/jellyfish/issues/252

File: website/docs/jellyfish/api/mining.md [hunk 5]
```diff
@@ -91,3 +91,25 @@ interface MasternodeInfo {
...
 }
 ```
+
+## estimateSmartFee
+
+Estimates the approximate fee per kilobyte needed for a transaction
+
+```ts title="client.mining.estimateSmartFee()"
+interface mining {
+  estimateSmartFee (confirmationTarget: number, estimateMode: EstimateMode = 'CONSERVATIVE'): Promise<SmartFeeEstimation>
```
Comment (fuxingloh) [inline website/docs/jellyfish/api/mining.md:101]:
```suggestion
  estimateSmartFee (confirmationTarget: number, estimateMode: EstimateMode = EstimateMode.CONSERVATIVE): Promise<SmartFeeEstimation>
```

File: packages/jellyfish-api-core/src/category/mining.ts [hunk 6]
```diff
@@ -37,6 +43,16 @@ export class Mining {
   async getMiningInfo (): Promise<MiningInfo> {
     return await this.client.call('getmininginfo', [], 'number')
   }
+
+  /**
+   *
+   * @param {number} confirmationTarget in blocks (1 - 1008)
+   * @param {EstimateMode} [estimateMode='CONSERVATIVE'] estimateMode of fees.
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/mining.ts:50]:
```suggestion
   * @param {EstimateMode} [estimateMode=EstimateMode.CONSERVATIVE] estimateMode of fees.
```

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/jellyfish-api-core/src/index.ts [hunk 7]
```diff
@@ -78,10 +78,10 @@ export class ClientApiError extends ApiError {
  * API RPC error, from upstream.
  */
 export class RpcApiError extends ApiError {
-  public readonly payload: { code: number, message: string }
+  public readonly payload: { code: number, message: string, rpcMethod?: string }
 
-  constructor (error: { code: number, message: string }) {
-    super(`RpcApiError: '${error.message}', code: ${error.code}`)
+  constructor (error: { code: number, message: string, rpcMethod?: string }) {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/index.ts:83]: Remove the optional aspect for `rpcMethod?: string` and add it to https://github.com/DeFiCh/jellyfish/blob/82b6692b48492bf109f95bbe3767f03bf32e0d7a/packages/jellyfish-api-jsonrpc/src/index.ts#L92-L94
as well

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/jellyfish-api-core/src/index.ts (hunk 7 above)
Comment (thedoublejay) [inline packages/jellyfish-api-core/src/index.ts:83]: @fuxingloh  @canonbrother I have added some changes on the `RpcApiError` class to make the rpcmethod visible on error logs.
image

Tests are running well locally but throwing errors on CI. Let me know if it's ok to merge, otherwise I can remove it when it's ready for review.

Comment (fuxingloh) [reply packages/jellyfish-api-core/src/index.ts:83]: I think varying CI conditions and CPU differences causes inconsistency in the test. 

Likely because the fees required to send increased hence the insufficient funds. To keep it consistent you can subtract fees from the amount. And with walletBalanceGTE() it will always have sufficient funds. By doing this you might need to increase the amount sent by x10.

https://github.com/DeFiCh/jellyfish/blob/82b6692b48492bf109f95bbe3767f03bf32e0d7a/packages/jellyfish-api-core/src/category/wallet.ts#L259

Comment (thedoublejay) [reply packages/jellyfish-api-core/src/index.ts:83]: Thanks! will try that

**Review by fuxingloh — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]
